components/layout/fragment.rs

```diff
@@ -2102,35 +2106,144 @@ impl Fragment {
         }
     }
 
-    pub fn strip_leading_whitespace_if_necessary(&mut self) {
-        let mut scanned_text_fragment_info = match self.specific {
+    pub fn strip_leading_whitespace_if_necessary(&mut self) -> WhitespaceStrippingResult {
+        if self.style.get_inheritedtext().white_space == white_space::T::pre {
+            return WhitespaceStrippingResult::RetainFragment
+        }
+
+        match self.specific {
             SpecificFragmentInfo::ScannedText(ref mut scanned_text_fragment_info) => {
-                scanned_text_fragment_info
+                let mut leading_whitespace_character_count = 0;
+                {
+                    let text = slice_chars(
+                        &*scanned_text_fragment_info.run.text,
+                        scanned_text_fragment_info.range.begin().to_usize(),
+                        scanned_text_fragment_info.range.end().to_usize());
+                    for character in text.chars() {
+                        if util::str::char_is_whitespace(character) {
+                            leading_whitespace_character_count += 1
+                        } else {
+                            break
+                        }
+                    }
+                }
+
+                let whitespace_range = Range::new(scanned_text_fragment_info.range.begin(),
+                                                  CharIndex(leading_whitespace_character_count));
+                let text_bounds =
+                    scanned_text_fragment_info.run.metrics_for_range(&whitespace_range).bounding_box;
+                self.border_box.size.inline = self.border_box.size.inline - text_bounds.size.width;
+                scanned_text_fragment_info.content_size.inline =
+                    scanned_text_fragment_info.content_size.inline - text_bounds.size.width;
+
+                scanned_text_fragment_info.range.adjust_by(
+                    CharIndex(leading_whitespace_character_count),
+                    -CharIndex(leading_whitespace_character_count));
+
+                return WhitespaceStrippingResult::RetainFragment
+            }
+            SpecificFragmentInfo::UnscannedText(ref mut unscanned_text_fragment_info) => {
+                let mut new_text_string = String::new();
+                let mut modified = false;
+                for (i, character) in unscanned_text_fragment_info.text.char_indices() {
+                    if gfx::text::util::is_bidi_control(character) {
```

Comment:
Should this use `is_always_discardable_char` (which in theory will also discard soft hyphens)?